### PR TITLE
Leverage `EvaluationContext` in `Cached` to avoid tripping execution-time access checks

### PR DESCRIPTION
**Pull request by abstratt**
Related to: #30860

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Review by mlopatkin — DISMISSED** [review]
Looks good!

:x: Please also move `EvaluationContextTest` and `EvaluationContextPerfTest`

File: platforms/core-runtime/functional/src/main/java/org/gradle/api/internal/provider/EvaluationContext.java
[the remark this answers is not in the document]
Comment (abstratt) [reply platforms/core-runtime/functional/src/main/java/org/gradle/api/internal/provider/EvaluationContext.java:1]: I wanted to minimize churn, as this was originally being done in #30886, which has a lot going on already. Now that is a separate PR, it does make sense.

File: platforms/core-runtime/functional/src/main/java/org/gradle/internal/evaluation/package-info.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Provides support for evaluating objects that offers features such as:
```
Comment (mlopatkin) [inline platforms/core-runtime/functional/src/main/java/org/gradle/internal/evaluation/package-info.java:18]: 👍

**Comment by abstratt** [comment]
`EvaluationContextPerfTest` uses `DefaultProperty` at the moment, so I left it in its original module, only moved it to the new package

**Review by mlopatkin — CHANGES_REQUESTED** [review]
Thanks!

One naming suggestion left. Sorry, it didn't occur to me when I was doing an initial pass.

**Comment by abstratt** [comment]
> One naming suggestion left. Sorry, it didn't occur to me when I was doing an initial pass.

No worries, thanks for the suggestions.

**Comment by abstratt** [comment]
@bot-gradle test PRF,ACC

**Pull request merged**